receiver/awscloudwatchreceiver/logs.go

```diff
@@ -0,0 +1,353 @@
+	FilterLogEventsWithContext(ctx context.Context, input *cloudwatchlogs.FilterLogEventsInput, opts ...request.Option) (*cloudwatchlogs.FilterLogEventsOutput, error)
+}
+
+type groupNames struct {
+	logGroup    string
+	streamNames []*string
+}
+
+func (gn *groupNames) request(limit int, nextToken string, st, et *time.Time) *cloudwatchlogs.FilterLogEventsInput {
+	base := &cloudwatchlogs.FilterLogEventsInput{
+		LogGroupName: &gn.logGroup,
+		StartTime:    aws.Int64(st.UnixMilli()),
+		EndTime:      aws.Int64(et.UnixMilli()),
+		Limit:        aws.Int64(int64(limit)),
+	}
+	if len(gn.streamNames) > 0 {
+		base.LogStreamNames = gn.streamNames
+	}
+	if nextToken != "" {
+		base.NextToken = aws.String(nextToken)
+	}
+	return base
+}
+
+type groupPrefix struct {
+	logGroup string
+	prefix   *string
+}
+
+func (gp *groupPrefix) request(limit int, nextToken string, st, et *time.Time) *cloudwatchlogs.FilterLogEventsInput {
+	base := &cloudwatchlogs.FilterLogEventsInput{
+		LogGroupName:        &gp.logGroup,
+		StartTime:           aws.Int64(st.UnixMilli()),
+		EndTime:             aws.Int64(et.UnixMilli()),
+		Limit:               aws.Int64(int64(limit)),
+		LogStreamNamePrefix: gp.prefix,
+	}
+	if nextToken != "" {
+		base.NextToken = aws.String(nextToken)
+	}
+	return base
+}
+
+type pollConfig interface {
+	request(limit int, nextToken string, st, et *time.Time) *cloudwatchlogs.FilterLogEventsInput
+}
+
+func newLogsReceiver(cfg *Config, logger *zap.Logger, consumer consumer.Logs) *logsReceiver {
+	namedPolls := []groupNames{}
+	prefixedPolls := []groupPrefix{}
+	for logGroupName, sc := range cfg.Logs.Groups.NamedConfigs {
+		for _, prefix := range sc.Prefixes {
+			prefixedPolls = append(prefixedPolls, groupPrefix{
+				logGroup: logGroupName,
+				prefix:   prefix,
+			})
+		}
+
+		namedPolls = append(namedPolls, groupNames{
+			logGroup:    logGroupName,
+			streamNames: sc.Names,
+		})
+	}
+
+	// safeguard from using both
+	autodiscover := cfg.Logs.Groups.AutodiscoverConfig
+	if len(cfg.Logs.Groups.NamedConfigs) > 0 {
+		autodiscover = nil
+	}
+
+	return &logsReceiver{
+		region:              cfg.Region,
+		profile:             cfg.Profile,
+		consumer:            consumer,
+		maxEventsPerRequest: cfg.Logs.MaxEventsPerRequest,
+		imdsEndpoint:        cfg.IMDSEndpoint,
+		autodiscover:        autodiscover,
+		pollInterval:        cfg.Logs.PollInterval,
+		namedPolls:          namedPolls,
+		prefixedPolls:       prefixedPolls,
+		logger:              logger,
+		wg:                  &sync.WaitGroup{},
+		doneChan:            make(chan bool),
+	}
+}
+
+func (l *logsReceiver) Start(ctx context.Context, host component.Host) error {
+	l.logger.Debug("starting to poll for Cloudwatch logs")
+
+	if l.autodiscover != nil {
+		namedPolls, prefixedPolls, err := l.discoverGroups(ctx, l.autodiscover)
+		if err != nil {
+			return fmt.Errorf("unable to auto discover log groups using prefix  %s: %w", l.autodiscover.Prefix, err)
+		}
```

Comment:
This will crash the collector, which should not happen if this fails.

Instead, we need to be resilient to transient issues. 

Even if this succeeded, as is we would never update the groups that are discovered. This is presumably not desired behavior either.